### PR TITLE
ScriptHost initialization refactor

### PR DESCRIPTION
**Pull request by mathewc**
These needed improvements are intended to simplify our init pipeline to lay the groundwork for more effective instrumentation and performance work. I'll merge this into v2 as well.

### CODE REVIEW

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 1]
```diff
@@ -84,6 +90,7 @@ public class ScriptHost : JobHost
             scriptConfig = scriptConfig ?? new ScriptHostConfiguration();
             _hostConfig = scriptConfig.HostConfig;
             _instanceId = Guid.NewGuid().ToString();
+            _storageConnectionString = AmbientConnectionStringProvider.Instance.GetConnectionString(ConnectionStringNames.Storage);
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Moving some readonly settings from init to ctor

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 2]
```diff
@@ -270,296 +281,27 @@ internal static void AddFunctionError(Dictionary<string, Collection<string>> fun
         }
 
         /// <summary>
-        /// Performs all required initialization on the host. Must be called before the host is started.
+        /// Performs all required initialization on the host.
+        /// Must be called before the host is started.
         /// </summary>
         public void Initialize()
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: I've moved all of the code out of this method, into pipeline stage functions that gather relatively cohesive operations. This method has went from about 300 lines to 12 lines :)

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 3]
```diff
@@ -270,296 +281,27 @@ internal static void AddFunctionError(Dictionary<string, Collection<string>> fun
...
-            FileUtility.EnsureDirectoryExists(hostLogPath);
-            string debugSentinelFileName = Path.Combine(hostLogPath, ScriptConstants.DebugSentinelFileName);
-            this.LastDebugNotify = File.GetLastWriteTime(debugSentinelFileName);
-
-            FunctionTraceWriterFactory = new FunctionTraceWriterFactory(ScriptConfig);
-
-            IMetricsLogger metricsLogger = CreateMetricsLogger();
-
-            using (metricsLogger.LatencyEvent(MetricEventNames.HostStartupLatency))
+            _stopwatch.Start();
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: I was using this stopwatch (and the logs it emits - see later below) during some local performance testing. Was going to remove, but now I wonder if we shouldn't just leave this in here. The logs are useful.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 4]
```diff
@@ -801,6 +543,314 @@ private void BlobLeaseManagerHasLeaseChanged(object sender, EventArgs e)
...
+        private void Preinitialize()
+        {
+            string debugSentinelFileName = Path.Combine(_hostLogPath, ScriptConstants.DebugSentinelFileName);
+            LastDebugNotify = File.GetLastWriteTime(debugSentinelFileName);
+
+            // take a startup time function directory snapshot so we can detect function additions/removals
+            // we'll also use this snapshot when reading function metadata as part of startup
+            // taking this snapshot once and reusing at various points during initialization allows us to
+            // minimize disk opoerations
+            _directorySnapshot = Directory.EnumerateDirectories(ScriptConfig.RootScriptPath).ToImmutableArray();
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: As part of some perf investigations, I'm now taking this snapshot early and reusing it in a couple places during init, which saves us 2 directory enumerations. I haven't measured the startup improvement from that yet, but on Azure Files I'd expect it to matter.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 5]
```diff
@@ -968,35 +1074,38 @@ private static FunctionMetadata ParseFunctionMetadata(string functionName, JObje
...
-            foreach (var scriptDir in Directory.EnumerateDirectories(config.RootScriptPath))
+            foreach (var scriptDir in functionDirectories)
             {
                 string functionName = null;
-
                 try
                 {
                     // read the function config
                     string functionConfigPath = Path.Combine(scriptDir, ScriptConstants.FunctionMetadataFileName);
-                    if (!File.Exists(functionConfigPath))
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Saved another file operation here, by optimistically attempting to read the file w/o the existence check

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 6]
```diff
@@ -1847,6 +1955,10 @@ internal static string GetAssemblyFileVersion(Assembly assembly)
 
         protected override void OnHostInitialized()
         {
+            string message = $"Host initialized ({_stopwatch.ElapsedMilliseconds}ms)";
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: This first log point is the earliest at which the host will accept http traffic. Listeners may not yet be started.

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 7]
```diff
@@ -1857,6 +1969,10 @@ protected override void OnHostStarted()
             HostStarted?.Invoke(this, EventArgs.Empty);
 
             base.OnHostStarted();
+
+            string message = $"Host started ({_stopwatch.ElapsedMilliseconds}ms)";
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Host is now fully started, including all listeners

File: src/WebJobs.Script/Diagnostics/MetricEventNames.cs [hunk 8]
```diff
@@ -14,7 +14,6 @@ public static class MetricEventNames
         public const string HostStartupInitializeBindingProvidersLatency = "host.startup.initializebindingproviders.latency";
         public const string HostStartupCreateMetadataProviderLatency = "host.startup.createmetadataprovider.latency";
         public const string HostStartupGetFunctionDescriptorsLatency = "host.startup.getfunctiondescriptors.latency";
-        public const string HostStartupPurgeLogDirectoriesLatency = "host.startup.purgelogdirectories.latency";
```
Comment (mathewc) [inline src/WebJobs.Script/Diagnostics/MetricEventNames.cs]: Now spinning up the purge operation on a background task, so it isn't a synchronous factor of the startup path anymore

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 9]
```diff
@@ -796,11 +538,319 @@ private void OnDebugModeFileChanged(object sender, FileSystemEventArgs e)
...
+        /// <summary>
+        /// Clean up any old files or directories.
+        /// </summary>
+        private void CleanupFileSystem()
+        {
+            if (ScriptConfig.FileLoggingMode != FileLoggingMode.Never)
+            {
+                // initiate the cleanup in a background task so we don't
+                // delay startup
+                Task.Run(() => PurgeOldLogDirectories());
```
Comment (mathewc) [inline src/WebJobs.Script/Host/ScriptHost.cs]: In the background now

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 10]
```diff
@@ -796,11 +538,319 @@ private void OnDebugModeFileChanged(object sender, FileSystemEventArgs e)
...
-        private void BlobLeaseManagerHasLeaseChanged(object sender, EventArgs e)
+        private void OnHostLeaseChanged(object sender, EventArgs e)
         {
             IsPrimaryChanged?.Invoke(this, EventArgs.Empty);
         }
 
+        /// <summary>
+        /// Perform any early initialization operations.
+        /// </summary>
+        private void Preinitialize()
```
Comment (paulbatum) [inline src/WebJobs.Script/Host/ScriptHost.cs]: Nit: should be pascal cased e.g. PreInitialize()

File: src/WebJobs.Script/Host/ScriptHost.cs [hunk 11]
```diff
@@ -796,11 +538,319 @@ private void OnDebugModeFileChanged(object sender, FileSystemEventArgs e)
...
+        {
+            string debugSentinelFileName = Path.Combine(_hostLogPath, ScriptConstants.DebugSentinelFileName);
+            LastDebugNotify = File.GetLastWriteTime(debugSentinelFileName);
+
+            // take a startup time function directory snapshot so we can detect function additions/removals
+            // we'll also use this snapshot when reading function metadata as part of startup
+            // taking this snapshot once and reusing at various points during initialization allows us to
+            // minimize disk opoerations
```
Comment (paulbatum) [inline src/WebJobs.Script/Host/ScriptHost.cs]: 'opoerations' typo

File: src/WebJobs.Script/Host/ScriptHost.cs (hunk 10 above)
Comment (mathewc) [reply src/WebJobs.Script/Host/ScriptHost.cs]: I knew it looked weird but couldn't figure out why...

**Pull request merged**